### PR TITLE
Refactor both DocumentBuilders to one DocumentFactory

### PR DESCRIPTION
**Pull request by JaZo**
## Description

I dropped both DocumentBuilders and created a `DocumentFactory` that makes an `ItemDocument` or `CollectionDocument` based on the data you give it. The factory doesn't hydrate the items anymore, that's up to the `ItemHydrator`. It does gather the included itself based on the relations of the item(s) instead of asking the item for the items that should be included.

## Motivation and context

This should improve the code quality by moving some logic to the responsible class and simplifying the `Item` (currently the worst rated class according to Scrutinizer).

## How has this been tested?

Tested with updated unit tests.

## Types of changes

What types of changes does your code introduce? Put an `x` in all the boxes that apply:
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

Go over all the following points, and put an `x` in all the boxes that apply.

- [x] I have read the **[CONTRIBUTING](CONTRIBUTING.md)** document.
- [x] My pull request addresses exactly one patch/feature.
- [x] I have created a branch for this patch/feature.
- [x] Each individual commit in the pull request is meaningful.
- [x] I have added tests to cover my changes.
- [x] If my change requires a change to the documentation, I have updated it accordingly.

### CODE REVIEW

File: src/DocumentFactory.php [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+     * @return \Swis\JsonApi\Client\Interfaces\DocumentInterface
+     */
+    public function make(DataInterface $data): DocumentInterface
+    {
+        $document = new Document();
+        if ($data instanceof ItemInterface) {
+            $document = new ItemDocument();
+        } elseif ($data instanceof Collection) {
+            $document = new CollectionDocument();
+        }
```
Comment (HendrikN) [inline src/DocumentFactory.php]: Wouldn't it be more appropriate to throw an exception here, when trying to _make_ an item of unknown type?

File: src/DocumentFactory.php [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+                }
+            )
+            ->flatMap(
+                function (ItemInterface $item) {
+                    return Collection::wrap($item)->merge($this->getIncludedFromItem($item));
+                }
+            )
+            ->filter(
+                static function (ItemInterface $item) {
+                    return $item->hasType() && $item->hasId() && ($item->hasAttributes() || $item->hasRelationships());
```
Comment (HendrikN) [inline src/DocumentFactory.php]: you could refactor this into a method (something like `itemCanBeIncluded`) for readability purposes

File: src/Item.php [hunk 3]
```diff
@@ -161,66 +161,12 @@ public function getRelationships(): array
...
-        if (null === $this->getId()) {
-            return false;
-        }
-
-        if (empty($this->relationships) && empty($this->toArray())) {
-            return false;
-        }
-
-        return true;
+        return !empty($this->getRelationships());
```
Comment (HendrikN) [inline src/Item.php]: Isnt empty a bit of an overkill here? Just `count($this->getRelationships())` would be enough.

File: src/Item.php [hunk 4]
```diff
@@ -261,6 +207,14 @@ public function hasAttribute($key): bool
         return array_key_exists($key, $this->attributes);
     }
 
+    /**
+     * @return bool
+     */
+    public function hasAttributes(): bool
+    {
+        return !empty($this->toArray());
```
Comment (HendrikN) [inline src/Item.php]: See previous comment

File: src/DocumentFactory.php (hunk 2 above)
Comment (JaZo) [reply src/DocumentFactory.php]: Better, thanks!

File: src/Item.php (hunk 3 above)
Comment (JaZo) [reply src/Item.php]: `empty` is generally considered faster/better than `count` for an array

**Pull request merged**